src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -51,78 +56,129 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
+      case true => ()
+    }
+  }
+
+  //Throws 404 and pretends we don't even know there's a cluster there, by default.
+  //If the cluster really exists and you're OK with the user knowing that, set throw401 = true.
+  protected def checkClusterPermission(user: UserInfo, action: NotebookClusterAction, cluster: Cluster, throw401: Boolean = false): Future[Unit] = {
+    authProvider.hasNotebookClusterPermission(user, action, cluster.googleId) map {
+      case false =>
+        if( throw401 )
+          throw AuthorizationError(user.userEmail)
+        else
+          throw ClusterNotFoundException(cluster.googleProject, cluster.clusterName)
+      case true => ()
     }
   }
 
-  def createCluster(serviceAccount: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
+  def createCluster(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
+    for {
+      _ <- checkProjectPermission(userInfo, CreateClusters, googleProject)
+      serviceAccount <- samDAO.getPetServiceAccount(userInfo)
+      cluster <- internalCreateCluster(userInfo.userEmail, serviceAccount, googleProject, clusterName, clusterRequest)
+    } yield cluster
+  }
+
+  def internalCreateCluster(userEmail: WorkbenchEmail, serviceAccount: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
     // Check if the google project has a cluster with the same name. If not, we can create it
     dbRef.inTransaction { dataAccess =>
       dataAccess.clusterQuery.getActiveClusterByName(googleProject, clusterName)
     } flatMap {
       case Some(_) => throw ClusterAlreadyExistsException(googleProject, clusterName)
       case None =>
         val augmentedClusterRequest = addClusterDefaultLabels(serviceAccount, googleProject, clusterName, clusterRequest)
-        val createFuture = for {
+        for {
           // Create the cluster in Google
-          (cluster, initBucket, serviceAccountKeyOpt) <- createGoogleCluster(serviceAccount, googleProject, clusterName, augmentedClusterRequest)
+          (cluster, initBucket, serviceAccountKeyOpt) <- createGoogleCluster(userEmail, serviceAccount, googleProject, clusterName, augmentedClusterRequest)
           // Save the cluster in the database
           savedCluster <- dbRef.inTransaction(_.clusterQuery.save(cluster, GcsPath(initBucket, GcsRelativePath("")), serviceAccountKeyOpt.map(_.id)))
-        } yield savedCluster
-
-        createFuture andThen {
-          case Success(cluster) =>
-            // Notify the cluster monitor upon cluster creation
-            clusterMonitorSupervisor ! ClusterCreated(cluster)
+          // Notify the auth provider that the cluster has been created
+          _ <- authProvider.notifyClusterCreated(savedCluster.creator.value, savedCluster.googleProject.value, savedCluster.googleId)
```

Comment:
Do we want to notify the auth provider here, or from `ClusterMonitorActor` when the cluster is actually ready for use?

Reply:
We need to keep track of the resource ASAP, which is why I put it here.

This makes me think I should also cancel cluster creation if the provider says no, e.g. because there's already a resource assigned.

Reply:
oh, the create message is in the yield so this already happens. cool

Reply:
wait no it doesn't